### PR TITLE
Port serial port test sources

### PR DESCRIPTION
**Pull request by JeremyKuhne**
This ports the test code essentially as-is. These need converted to
xunit based tests and included in the test project as converted.

I'll convert something in a subsequent check-in to set the example.

@ianhays, @karelz, @willdean, @msalsbery, @SWY1985 

#984

### CODE REVIEW

File: src/System.IO.Ports/tests/Legacy/SerialPort/SerialPortRegressions.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            return false;
+        }
+
+        return true;
+    }
+
+
+    public bool VerifyReadExisting(System.Text.Encoding encoding)
+    {
+        string text = "????????????4??????????????????,?11????????????????????????????????????????????????,????????????,??????????";
```
Comment (willdean) [inline src/System.IO.Ports/tests/Legacy/SerialPort/SerialPortRegressions.cs]: Is this string something in unicode of some kind that hasn't come across right, or is that just GitHub?

File: src/System.IO.Ports/tests/Legacy/SerialPort/USB_Serial_Regression.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// This is a regression test for Dev10 #511456, DevDiv Bugs #186131, and DevDiv Bugs #191753.
+// Unplugged USB virtual serial port causes SerialPort object to crash application in finalizer thread.
+//
+// For this manual test, you will need a USB Serial adapater, probably one like this:
```
Comment (willdean) [inline src/System.IO.Ports/tests/Legacy/SerialPort/USB_Serial_Regression.cs]: Are we going to try and retain these sort of manual tests within CoreFx?

File: src/System.IO.Ports/tests/Legacy/SerialPort/BaseStream.cs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+        if (com1.IsOpen)
+            com1.Close();
+
+        return retValue;
+    }
+
+    public bool BaseStream_Open()
+    {
+        SerialPort com1 = new SerialPort(TCSupport.LocalMachineSerialInfo.FirstAvailablePortName);
```
Comment (willdean) [inline src/System.IO.Ports/tests/Legacy/SerialPort/BaseStream.cs]: Is TCSupport something we have?

**Comment by willdean** [comment]
This is a veritable cathedral of serial port testing, I'm impressed.

It does look to me like it mostly relies on real hardware to test against, even where that isn't explicitly called-out.  Presumably it was documented somewhere what the h/w requirements were to run these tests - if so, can that document be brought across / published somewhere?

File: src/System.IO.Ports/tests/Legacy/SerialPort/SerialPortRegressions.cs (hunk 1 above)
Comment (danmosemsft) [inline src/System.IO.Ports/tests/Legacy/SerialPort/SerialPortRegressions.cs]: Oops, you're right. Needs to be converted to \unnnn

File: src/System.IO.Ports/tests/Legacy/SerialPort/BaseStream.cs (hunk 3 above)
Comment (willdean) [inline src/System.IO.Ports/tests/Legacy/SerialPort/BaseStream.cs]: Sorry, found it now.

File: src/System.IO.Ports/tests/Legacy/SerialPort/USB_Serial_Regression.cs (hunk 2 above)
Comment (JeremyKuhne) [inline src/System.IO.Ports/tests/Legacy/SerialPort/USB_Serial_Regression.cs]: Yes, we have other cases of manual tests floating around other projects too (Console comes to mind). In this case, adding a skip to the attribute that you have to manually remove.

**Comment by JeremyKuhne** [comment]
> It does look to me like it mostly relies on real hardware to test against, even where that isn't explicitly called-out. Presumably it was documented somewhere what the h/w requirements were to run these tests - if so, can that document be brought across / published somewhere?

If I find anything I'll bring it. As you can see these tests were mostly written over 13 years ago- finding documentation can be really tough as we went through many iterations of various collaborative tools (wikis, various Sharepoint versions, etc., etc.). Be pleasantly surprised if I find something. :)

In any case, we need to set up a base class for the tests that we can use to provide properties to condition the tests based on specific hardware needs using `[ConditionalFact()]`. I'm going to rely on your input to try and find a workable solution for configuring the test machines and we can look to get CI and/or outer loop hardware set up correctly.

File: src/System.IO.Ports/tests/Legacy/SerialPort/SerialPortRegressions.cs (hunk 1 above)
Comment (JeremyKuhne) [inline src/System.IO.Ports/tests/Legacy/SerialPort/SerialPortRegressions.cs]: Thanks. I'll follow up with my next change.

**Comment by willdean** [comment]
I haven't gone through it in detail (only looking in GitHub at moment), but it's clear that tests are already able to define what they need in the way of hardware, and then they're run or not run depending on what specific hardware's available -  we'll just have to work out a list of what runs for each hardware configuration, and then see if we're happy with any of the subsets.

**Comment by karelz** [comment]
@JeremyKuhne I happened to clean up most of our internal SharePoints and wikis over the last few years and migrate them to http://clr - just search there. If it is not there, it is lost.

**Comment by msalsbery** [comment]
Excellent! I volunteered to contribute to the tests/testing so I'll start digging in, after I check out the contributor guides since I'm brand new here

**Pull request merged**